internal/hcsoci/pnp.go

```diff
@@ -0,0 +1,58 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	"github.com/pkg/errors"
+)
+
+const uvmPnpExePath = "C:\\Windows\\System32\\pnputil.exe"
+
+// createPnPInstallDriverCommand creates a pnputil command to add and install drivers
+// present in `driverUVMPath` and all subdirectories.
+func createPnPInstallDriverCommand(driverUVMPath string) []string {
+	dirFormatted := fmt.Sprintf("%s/*.inf", driverUVMPath)
+	args := []string{
+		"cmd",
+		"/c",
+		uvmPnpExePath,
+		"/add-driver",
+		dirFormatted,
+		"/subdirs",
+		"/install",
+		"&",
+	}
+	return args
+}
+
+// createPnPInstallAllDriverArgs creates a single command to add and install
+// a set of driver locations in the UVM
+func createPnPInstallAllDriverArgs(driverUVMDirs []string) []string {
+	var result []string
+	for _, d := range driverUVMDirs {
+		driverArgs := createPnPInstallDriverCommand(d)
+		result = append(result, driverArgs...)
+	}
+	return result
+}
+
+// execPnPInstallAllDrivers makes the call to exec in the uvm the pnp command
+// that installs all drivers previously mounted into the uvm.
+func execPnPInstallAllDrivers(ctx context.Context, vm *uvm.UtilityVM, driverDirs []string) error {
```

Comment:
Should we track the added drivers and remove them when the container exits?

Reply:
I think that is a plan for the next iteration on top of this, so I was planning on putting it in a different PR. There's not an explicit ask for it right now, though I think we definitely need it for scenarios where successive containers that don't overlap in lifetime use descending driver versions.